src/library_manager/lib_notification.c

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Jaroslaw Stelter <jaroslaw.stelter@intel.com>
+//         Pawel Dobrowolski <pawelx.dobrowolski@intel.com>
+//
+
+/*
+ * 3-rd party modules do not have access to component device runtime data.
+ * System service API used by these modules require notification handling
+ * based on base FW message management. Therefore this code expose notification
+ * calls aligned with cAVS/ACE System service API requirements.
+ */
+#include <sof/sof.h>
+#include <sof/list.h>
+#include <sof/ipc/msg.h>
+#include <sof/lib/memory.h>
+#include <sof/lib_manager.h>
+
+LOG_MODULE_DECLARE(lib_manager, LOG_LEVEL_INFO);
+
+struct ipc_msg *lib_notif_msg_init(uint32_t header, uint32_t size)
+{
+	struct ipc_msg *msg = NULL;
+	struct ipc_lib_msg *msg_pool_elem;
+	struct ext_library *ext_lib = ext_lib_get();
+	struct ipc_lib_msg *lib_notif = ext_lib->lib_notif_pool;
```

Comment:
your `.lib_notif_pool` is somewhat... unusual. Usually you'd just have a list head in the higher level object (in `ext_lib` in this case) to which you then link elements as you allocate them. Instead you have a pointer to `lib_notif_pool` there and once you allocate the first element and attach it there, you then *link* further allocated elements to the first one. That makes the list head instance in that object both - a list head and a list element. `list_empty()` now means not 0 elements, but 1... All that can be made to work of course but just is a bit unusual and one has to remember about this when working with this code.

Reply:
So what would you suggest. That might be refactored for sure but I think that main goal is to port that feature from legacy and integrate. Many mechanisms used here are borrowed from older firmware and for sure they can be improved. We can do that later if it isn't so wrong.